Update CCNode.cpp

dadidzf opened the pull request:
Retain the node before update for lua, in case the lua update remove self and cause crashes !

Example(lua):
node = cc.Node:create()
node:onUpdate(handler(self, self.nodeUpdate))
node.nodeUpdate = function(self, dt)
    self:removeFromParent()             -- Cause crashes, _componentContainer has been a wild pointer !
end

Crash Point:
void Node::update(float fDelta)
{
#if CC_ENABLE_SCRIPT_BINDING
if (0 != _updateScriptHandler)
    {
    //only lua use
    SchedulerScriptData data(_updateScriptHandler,fDelta);
    ScriptEvent event(kScheduleEvent,&data);
    ScriptEngineManager::getInstance()->getScriptEngine()->sendEvent(&event);
}
#endif
    
    if (_componentContainer && !_componentContainer->isEmpty())  -- Crash
    {
        _componentContainer->visit(fDelta);
    }
}

Commit:
Update CCNode.cpp

Retain the node before update for lua, and release it after

Dauch commented:
you need to add to a parent to make it retain, not modify the CCNode.cpp do do the change.

dumganhar commented on cocos/2d/CCNode.cpp:
```diff
@@ -1630,6 +1630,7 @@ void Node::pauseSchedulerAndActions()
 // override me
 void Node::update(float fDelta)
 {
+    this->retain();
```
Is it better to put retain/release things into `CC_ENABLE_SCRIPT_BINDING` area?

dadidzf replied:
Yes, It's better ! ^_^

Commit:
Update CCNode.cpp

Put the retain/release inside CC_ENABLE_SCRIPT_BINDING area

Commit:
Update CCNode.cpp

Adjust retain/release position, inside the CC_ENABLE_SCRIPT_BINDING area will make no effect

Commit:
Update CCNode.cpp

dadidzf commented:
I did not find a good way to add the retain/release pairs, so close it at present
